pkg/time/exponent-ticker.go

```diff
@@ -0,0 +1,120 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package time
+
+import (
+	"math"
+	"time"
+)
+
+// ExponentTicker - holds a channel that delivers `ticks' of a clock at exponential
+// interval of minimum duration but not greater than maximum duration.
+type ExponentTicker struct {
+	timer  *time.Timer
+	doneCh chan struct{}
+	C      <-chan time.Time
+}
+
+// Stop - turns off a ticker. After Stop, no more ticks will be sent. Stop closes the
+// channel hence a read from the channel needs to check whether the channel is closed
+// or not.
+func (ticker *ExponentTicker) Stop() {
+	if ticker.timer != nil {
+		ticker.timer.Stop()
+	}
+
+	if ticker.doneCh != nil {
+		close(ticker.doneCh)
+	}
+}
+
+// NewExponentTicker - returns a new ExponentTicker containing a channel that
+// will send `ticks' of a clock at exponential interval of minimum duration
+// (min * 2 ** x) but not greater than maximum duration.  min and max
+// durations must be greater than zero and max duration must not be less than
+// min duration; if not, NewExponentTicker will panic.
+// Stop the ticker to release associated resources.
+func NewExponentTicker(min, max time.Duration) *ExponentTicker {
+	if min <= time.Duration(0) {
+		panic("non-positive min duration for NewExponentTicker")
+	}
+
+	if max <= time.Duration(0) {
+		panic("non-positive max duration for NewExponentTicker")
+	}
+
+	if min > max {
+		panic("min duration must be greater than max duration for NewExponentTicker")
+	}
+
+	var exponent float64
+	getExponentDuration := func() time.Duration {
+		d := min * time.Duration(math.Exp2(exponent))
+		if d > max {
+			d = min
```

Comment:
Not sure exactly about what's the purpose of resetting values here: I mean if the last operation fails, then it would make sense to continue with the maximum timeout.

Reply:
My understanding on the requirement is to reset once make it reached max duration.  If I am wrong and behavior the one you described, please confirm.